### PR TITLE
reconciler: support disconnected clients

### PR DESCRIPTION
**Pull request by DerekStrickland**
**Description**

This PR is a work in progress tracking against the long-running feature branch `f-disconnected-client-handling`.

It includes:

- New node status (`NodeStatusDisconnected`), new client status (`AllocClientStatusUnknown`), and a new eval trigger (`EvalTriggerMaxDisconnectTimeout`)
- Updates to the `taintedNodes` function to include disconnected nodes as part of the tainted set
- A new `MaxClientDisconnect` field on the `Allocation` struct for tracking how long an alloc is allowed to be disconnected
- `allocReconciler` changes to support allocs that can disconnect and reconnect without restarting
- `GenericScheduler` changes to add disconnects to the `NodeAllocation` set for upsert with new status `AllocClientStatusUnknown`
- `GenericScheduler` changes to add reconnects to `queuedAllocations`.
- Changes to the `StateStore` alloc upsert code so that allocs with the unknown status don't have their `ClientStatus` overwritten by the existing allocs `ClientStatus`

### CODE REVIEW

File: nomad/structs/structs.go
Comment (DerekStrickland) [inline nomad/structs/structs.go:10958]: Needs review.

File: scheduler/util.go
Comment (DerekStrickland) [inline scheduler/util.go:358]: Fixed in later commit.

File: client/client.go
Comment (tgross) [inline client/client.go:2414]: This is already a very large PR. Does client code really belong in it?

File: scheduler/generic_sched_test.go
Comment (tgross) [inline scheduler/generic_sched_test.go:6519]: What's going on here?

File: scheduler/generic_sched_test.go
Comment (tgross) [inline scheduler/generic_sched_test.go:6398]: If we're going to do table tests, provide a name so that it gives us readable test output:

```suggestion
		t.Run(tc.name, func(t *testing.T) {
```

There's currently only one test here though and the test assertions look pretty complex. Maybe not worth having a table?

**Review by tgross — COMMENTED** [review]
Hi @DerekStrickland! This is looking great for a first pass at some gnarly scheduler code. A bunch of my comments are about docstrings, readability, testing, etc. 
I've got a couple of correctness confusions around:

* How we're using `MaxNormScore()` in `computeStopByReconnecting`
* Confusing conditionals in `filterByTainted` when tainted nodes are nil that might be obscuring the right behavior
* Batch creation in `createTimeoutLaterEvals`

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:29]: This was my IDE "helping" me. I've disabled this feature and will get things patched up in the morning.

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6398]: I'm hoping to add more in later rounds. I can delete this table, but then I will just have to add it back later as I flesh out tests. Is it ok to leave as is for now?

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6486]: ooops. yes.

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6519]: I used another test as a basis for this and was unsure if I would need to append `AllocStates` so I left it as a reminder. I should have added a `TODO (derek)` tag with some comments. Sorry for that!

Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6519]: Or maybe opening an issue is the right workflow?  As you've seen previously, my habit from `$lastJobs` is to create `TODO` comments. I could use some input so I don't create more cruft for you to have to filter out. Issues feel kind of heavy-handed, but maybe that's preferred? So I'm not sure.

File: scheduler/reconcile.go
Comment (DerekStrickland) [inline scheduler/reconcile.go:1304]:
```suggestion
// appendFollowupEvals appends a set of followup evals for a task group to the
```

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6519]: I see now in a later comment that the preferred method is to move todos to a new branch. I'll follow that advice.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:29]: Fixed in next push.

File: scheduler/generic_sched_test.go
Comment (tgross) [reply scheduler/generic_sched_test.go:6519]: > Or maybe opening an issue is the right workflow? As you've seen previously, my habit from $lastJobs is to create TODO comments. I could use some input so I don't create more cruft for you to have to filter out. Issues feel kind of heavy-handed, but maybe that's preferred? So I'm not sure.

The trouble with TODOs is that they tend to rot in the code base. But this PR is also targeting a feature branch where you could potentially come by and cleanup your TODOs before they land in `main`. So having a TODO here without a corresponding issue is fine so long as it doesn't end up landing in `main` like that. (I would also recommend adding a checklist item to your umbrella issue in team-nomad to make it harder to forget, but that's because I'm super forgetful if I don't use external memory 😀 ). Then once things are getting ready to land in `main` and will live "forever", writing up a quick issue for the TODO will give it a shot at getting done some day.

> I see now in a later comment that the preferred method is to move todos to a new branch. I'll follow that advice.

If you're referring to [this](https://github.com/hashicorp/nomad/pull/12058#discussion_r806147723) comment, what I'm saying there is not to add more stuff onto this already-large PR. Instead, come back and implement it as a new PR so that we can get this one merged and reduce the amount of rebasing you need to do in parallel branches.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1068]: Ok. After re-reading all of this, it looks like `PopulateScoreMetaData` has already handled all the sorting of scores, so `a.ScoreMetaData[0]` should have the highest score. My next push will have this update in it. I'll let you know when it's pushed.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:10958]: I'm not sure if I need to set the `Job` and `Resources` `nil` here. I modeled this after `AppendStoppedAlloc`, but I don't understand why these are set to `nil`.

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go]: Removing until I can find a way to table test.

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go:6519]: Removed in next push.

File: scheduler/reconcile_test.go
Comment (DerekStrickland) [reply scheduler/reconcile_test.go:5081]: Removed in next push.

File: scheduler/reconcile_util.go
Comment (DerekStrickland) [reply scheduler/reconcile_util.go:251]: Yes, we do. The comments indicate we do this to handle nodes that have been GC'd.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:434]: Updated in the next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1155]: Agreed, we should not. I meant to remove these logs too. I'm removing this last conditional and the logs. Will be available in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1145]: Log entries removed so language removed.

Comment (DerekStrickland) [reply scheduler/reconcile.go:1145]: Fixed in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1003]: I have 100% been mentally modeling this all wrong. I've been optimizing function comments for code maintainers rather than doc strings. 

Updated in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1025]: Yeah. I was trying to reduce iterations. I've combined the conditionals as you suggested, which is a clear improvement. If that still feels like too much after you see the update, let me know and I can break it up further. I also added a guard clause if remove == 0 and changed these `logger.Debug` to `logger.Error` since I think these are actual error conditions.

Updated in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1068]: I think now that I've refactored `MaxNormScore` we are using the `FinalScore` calculated in `rank.go`. If there is still uncertainty, maybe we could look at it synchronously. It's a bit tricky 😬

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1229]: Updated in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1270]: Yes. Fixed in next push.

File: scheduler/reconcile.go
Comment (DerekStrickland) [reply scheduler/reconcile.go:1271]: This was 100% related to the previous copy-paste error. 

Fixed in the next push. Leaving this comment open in case it still looks wrong when you review it.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:10944]: Thanks for the guidance on where to look. It seems like this is an attempt to reduce the payload size as it serializes a `ApplyPlanResultRequest`. There is one Job on the whole request, and the `NodeAllocations` are all added, so stripping these fields should significantly reduce the size.

File: scheduler/reconcile_util.go
Comment (tgross) [reply scheduler/reconcile_util.go:251]: Ok, makes sense. Not having strict data integrity around GC bites us with having to deal with `nil` objects all the time (ex. #11931) but there's nothing we should be doing about it in this project for sure. 👍

File: scheduler/reconcile_util.go
Comment (tgross) [reply scheduler/reconcile_util.go:256]: This question was resolved but I don't see anything that addresses it?

File: nomad/structs/structs.go
Comment (tgross) [inline nomad/structs/structs.go:9735]: Can `a.Job` ever be nil? I know we set it that way in some circumstances.

File: scheduler/reconcile.go
Comment (tgross) [inline scheduler/reconcile.go:1128]:
```suggestion
		// If the user has defined a DesiredTransition don't resume the alloc.
```

File: scheduler/reconcile.go
Comment (tgross) [inline scheduler/reconcile.go:1133]:
```suggestion
		// If the scheduler has defined a terminal DesiredStatus don't resume the alloc.
```

File: scheduler/reconcile.go
Comment (tgross) [reply scheduler/reconcile.go]: Yeah that reads better 👍

File: scheduler/reconcile.go
Comment (tgross) [reply scheduler/reconcile.go:1053]: Ah, I see. :+1:

File: nomad/structs/structs.go
Comment (tgross) [reply nomad/structs/structs.go]: 👍

**Review by tgross — APPROVED** [review]
LGTM!

I've left a few minor remaining questions or suggestions but once those are resolved, let's get this merged.

File: nomad/structs/structs.go
Comment (DerekStrickland) [reply nomad/structs/structs.go:9735]: Updated in next push.

File: scheduler/reconcile_util.go
Comment (DerekStrickland) [reply scheduler/reconcile_util.go:256]: I've added a clause that places pending allocs on disconnected nodes into the lost set. Not sure what happened with that previously.

File: scheduler/generic_sched_test.go
Comment (DerekStrickland) [reply scheduler/generic_sched_test.go]: I was able to remove the second one, but not the first. If I remove the first the test will fail.  My belief is that since Process on the scheduler calls the TestHarness's SubmitPlan implementation which does a lot of things, the eval just isn't in the StateStore by then. I modeled this after other tests in this file that perform this same Wait, so maybe it is known?  This is the last item that I am aware of, so once you approve, I'll merge.

Comment (tgross) [reply scheduler/generic_sched_test.go]: Hm, that seems like a bug in the test harness then unless I'm _really_ misunderstanding something. I'll dig into what the test harness is assuming when I have some free time. Let's not hang up merging on that.

**Pull request merged**
